### DISCUSSION TITLE
Any way to declare a website is compatible with dark mode and no need to inject style again?

### DISCUSSION THREAD

**Discussion by dreamerblue** [darkreader/darkreader#9802] [General]
Many of our websites have designed light & dark theme and we don't wanna DarkReader to inject dark style it again, it will break our design.

I have referenced #1114 and tried to add meta in `<head>` but DarkReader is still injecting style:

```
<meta name="color-scheme" content="dark light">
<meta name="darkreader" content="uuid">
```

I think DarkReader should let developer to declare they have implemented dark style and stop injection. We prefer this way because submitting PR to `dark-sites.config` need to wait a lot (users cannot get new data rapidly) and it will break and delay our release plan.

**Comment by Gusted** [comment]
Please check #4325

**Reply by dreamerblue** [reply]
Thanks, I have found a PR trying to solve this (https://github.com/darkreader/darkreader/pull/9181), but it only works on switching (ON/OFF), not page on load. So as you said, the first detection is happened before <head> loaded...
I will keep focus on progress if there are any way to do that. Now i may try to add a listener to remove injected code by myself :(

**Reply by Gusted** [reply]
> , but it only works on switching (ON/OFF), not page on load. So as you said, the first detection is happened before loaded...

That shouldn't happen. The code passively listens for this element to be added to the DOM.

**Reply by dreamerblue** [reply]
I tested such html template below on Edge and Chrome. If this doesn't work as expected, I may open an issue.

```html
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1,minimum-scale=1,user-scalable=no">
  <meta name="color-scheme" content="dark light">
  <meta name="darkreader-lock">
  <title>Title</title>
  <style>body{margin:0;background-color:#f0f2f5;}@media(prefers-color-scheme:dark){body{background-color:#000;}}</style>
  <script>
    console.log('check whether darkreader injected style:', document.head.querySelector('.darkreader'));
    var systemThemeMediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
    document.documentElement.className = systemThemeMediaQuery.matches ? 'dark' : 'light';
  </script>
</head>

<body>
  <div id="root"></div>
</body>

</html>
```

Output:
<img width="850" alt="image" src="https://user-images.githubusercontent.com/19774268/189560050-d673cc8c-560c-4b65-a0b3-0d288c83bb9e.png">

**Reply by Gusted** [reply]
> I tested such html template below on Edge and Chrome. If this doesn't work as expected, I may open an issue.

I'm not sure how this doesn't work as expected, are you actually verifying the results by looking at the site itself?

Due to how the DOM is loading and Dark Reader trying to inject as early as possible, it might be a few moments that there is a darkreader element in the DOM. Which is what the code detects.

Try this modified example with a small timeout:
```html
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1,minimum-scale=1,user-scalable=no">
  <meta name="color-scheme" content="dark light">
  <meta name="darkreader-lock">
  <title>Title</title>
  <style>body{margin:0;background-color:#f0f2f5;}@media(prefers-color-scheme:dark){body{background-color:#000;}}</style>
  <script>
    setTimeout(() => {
    console.log('check whether darkreader injected style:', document.head.querySelector('.darkreader'));
    }, 5);
    var systemThemeMediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
    document.documentElement.className = systemThemeMediaQuery.matches ? 'dark' : 'light';
  </script>
</head>

<body>
  <div id="root"></div>
</body>

</html>
```

**Reply by dreamerblue** [reply]
Yes, I tried and the output is same as before. The injection is really early so that it even cannot detect any meta tag.

**Reply by dreamerblue** [reply]
like this example:
```html
<!DOCTYPE html>
<html lang="en">

<head>
  <script>
    console.log('check meta:', document.head.querySelector('meta[name=darkreader-lock]'));
    console.log('check whether darkreader injected style:', document.head.querySelector('.darkreader'));
    setTimeout(() => {
      console.log('check whether darkreader injected style after timeout:', document.head.querySelector('.darkreader'));
    }, 5);
  </script>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1,minimum-scale=1,user-scalable=no">
  <meta name="color-scheme" content="dark light">
  <meta name="darkreader-lock">
  <title>Title</title>
  <style>body{margin:0;background-color:#f0f2f5;}@media(prefers-color-scheme:dark){body{background-color:#000;}}</style>
  <script>
    var systemThemeMediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
    document.documentElement.className = systemThemeMediaQuery.matches ? 'dark' : 'light';
  </script>
</head>

<body>
  <div id="root"></div>
</body>

</html>
```

The output is:
<img width="693" alt="image" src="https://user-images.githubusercontent.com/19774268/190184723-9d682f49-5df5-4406-bd8a-80374adb1103.png">

It looks the injection happens really early, I am not sure how it did this.

**Reply by Gusted** [reply]
Yes that's correct, there's always a injection into a site very early on, this is to prevent #1219. But what has this to do with the original question? It's quite simple, use `<meta name="darkreader-lock">` and Dark Reader will not inject itself.

**Reply by dreamerblue** [reply]
Oh, yes. The issue is I cannot simply declare "My website has developed dark mode and it's no need to inject style again" by using `<meta name="darkreader-lock">` (This meta tag only works when switching on/off on Dark Reader popup window). Dark Reader will inject some styles before checking <meta>, so i can only add a script to manually remove injected code:

```html
<script>
      const __darkReaderInjected = document.head.querySelector('.darkreader');
      __darkReaderInjected && __darkReaderInjected.remove();
</script>
```

**Reply by dreamerblue** [reply]
Now back to the original question, my current solution to declare that a website is compatible with dark mode:
1. add `<meta name="darkreader-lock">`
2. add script to remove injected style (`.darkreader`) manually

Is there any more graceful way to do this?

**Reply by Gusted** [reply]
There was a bug, it wouldn't always remove the injected style. This will be fixed in the next version: #9914 so from the next version onwards you will just need step 1.

**Reply by dreamerblue** [reply]
Thanks!
